File: lib/geoscript/feature/feature.js
var register = require("./util").register;
var Factory = require("../factory").Factory;
var GEOM = require("../geom");
var jts = Packages.com.vividsolutions.jts;
var SimpleFeatureBuilder = Packages.org.geotools.feature.simple.SimpleFeatureBuilder;
var UTIL = require("../util");
var NUTIL = require("util");
var Schema = require("./schema").Schema;

var prepConfig = function(config) {
    config = UTIL.apply({}, config);
    if (!config.schema) {
        if (config.values) {
            config.schema = Schema.fromValues(config.values);
        }
    }
    return config;
};

/** api: (define)
 *  module = feature
 *  class = Feature
 */
var Feature = UTIL.extend(Object, {
    
    /** private: property[cache]
     *  ``Object``
     *  Cache for setters & getters.
     */ 
    cache: null,
    
    /** api: property[schema]
     *  :class:`feature.Schema`
     *  The feature schema.
     */
    schema: null,
    
    /** api: constructor
     *  .. class:: Feature
     *  
     *      :arg config: ``Object`` Configuration object.
     *
     *      Create a new feature.
     */
    constructor: function Feature(config) {
        this.cache = {};
        if (config) {
            config = prepConfig(config);
            if (config.schema) {
                this.schema = config.schema;                
            } else {
                throw "Feature must be constructed with a schema or values.";
            }

            var builder = new SimpleFeatureBuilder(this.schema._schema);
            var value, defaultGeom;
            for (var name in config.values) {
                value = config.values[name];
                if (value instanceof GEOM.Geometry) {
                    if (!defaultGeom) {
                        defaultGeom = value;
                    }
                    value = value._geometry;
                }
                builder.set(name, value);
            }
            if (defaultGeom) {
                // cache the default geometry
                this.cache.geometry = defaultGeom;
            }
            this._feature = builder.buildFeature(config.id); 
        }
    },    
    
    /** api: property[id]
     *  ``String``
     *  The feature identifier.  Read only.
     */
    get id() {
        return String(this._feature.identifier);
    },
    
    /** api: property[geometryName]
     *  ``String``
     *  Field name for the default geoemtry, or ``undefined`` if the feature
     *  has no geometry.
     */
    get geometryName() {
        var name;
        var geomProp = this._feature.getDefaultGeometryProperty();
        if (geomProp) {
            name = String(geomProp.getName());
        }
        return name;
    },
    
    /** api: property[geometry]
     *  :class:`geom.Geometry`
     *  The default geometry (if any) for the feature.  Will be ``undefined`` 
     *  if the feature does not have a geometry.
     */
    get geometry() {
        if (this.cache.geometry == null) {
            var _geometry = this._feature.getDefaultGeometry();
            if (_geometry) {
                this.cache.geometry = GEOM.Geometry.from_(_geometry); // may be undefined
            }
        }
        return this.cache.geometry;
    },
    set geometry(geometry) {
        this._feature.setDefaultGeometry(geometry._geometry);
        this.cache.geometry = geometry;            
    },
    
    /** api: property[projection]
     *  :class:`proj.Projection`
     *  Optional projection for the feature.  This corresponds to the projection
     *  of the default geometry for the feature.
     */
    get projection() {
        return this.geometry && this.geometry.projection;
    },
    set projection(projection) {
        if (this.geometry) {
            this.geometry.projection = projection;
        }
    },
    
    /** api: property[bounds]
     *  :class:`geom.Bounds`
     *  The bounds of the default geometry (if any) for this feature.  Will be
     *  ``undefined`` if the feature has no geometry.
     */
    get bounds() {
        var bounds;
        if (this.geometry) {
            return this.geometry.bounds;
        }
    },
    
    /** api: method[set]
     *  :arg name: ``String`` Attribute name.
     *  :arg value: ``String`` Attribute value.
     *
     *  Set a feature attribute.
     */
    set: function(name, value) {
        var field = this.schema.get("name");
        if (!field) {
            throw "Feature schema has no field named '" + name + "'";
        }
        if (value instanceof GEOM.Geometry) {
            if (name === this.geometryName) {
                this.geometry = value;
            }
            value = value._geometry;
        }
        this._feature.setAttribute(name, value);
    },
    
    /** api: method[get]
     *  :arg name: ``String`` Attribute name.
     *
     *  Get an attribute value.
     */
    get: function(name) {
        var value;
        if (name === this.geometryName) {
            value = this.geometry;
        } else {
            value = this._feature.getAttribute(name);
            if (value instanceof jts.geom.Geometry) {
                value = GEOM.Geometry.from_(value);
            } else if (typeof value !== "string" && typeof value !== "number") {
                if (value instanceof java.lang.Number) {
                    value = Number(value);
                } else {
                    value = String(value);
                }
            }
        }
        return value;
    },
    
    /** api: property[values]
     *  ``Object``
     *  An object with all the feature property names and values.  Used for 
     *  property access only.  Use :meth:`set` to set property values.
     */
    get values() {
        var values = {};
        if (this.schema) {
            this.schema.fieldNames.forEach(function(name) {
                values[name] = this.get(name);
            }, this);            
        }
        return values;
    },

    /** api: property[json]
     *  ``String``
     *  The JSON representation of the feature (see http://geojson.org).
     */
    get json() {
        return JSON.encode(this.config);
    },
    
    /** private: property[config]
     *  ``Object``
     */
    get config() {
        var values = this.values;
        var properties = {}, val;
        for (var name in values) {
            val = values[name];
            if (typeof val === "number" ||
                typeof val === "string" ||
                val === null) {
                properties[name] = val;
            }
        }
        return {
            type: "Feature",
            geometry: this.geometry && this.geometry.config || null,
            properties: properties
        };
    },

    /** private: method[toFullString]
     */
    toFullString: function() {
        var values = this.values;
        var value;
        var items = [];
        for (var name in values) {
            value = values[name];
            if (value instanceof GEOM.Geometry) {
                items.push(name + ": <" + value.constructor.name + ">");
            } else {
                items.push(name + ": " + NUTIL.repr(value));
            }
        }
        return items.join(", ");
    }
    
});

/** private: staticmethod[from_]
 *  :arg _feature: ``org.geotools.feature.simple.SimpleFeatureImpl``
 *  :arg _schema: ``org.geotools.feature.simple.SimpleFeatureTypeImpl`` Optional schema.
 *  :returns: :class:`Feature`
 *
 *  Create a feature from corresponding GeoTools parts.
 */
Feature.from_ = function(_feature, _schema) {
    var feature = new Feature();
    feature._feature = _feature;
    feature.schema = Schema.from_(_schema || _feature.type);
    return feature;
};

/** api: example
 *  Sample code to create a new feature:
 * 
 *  .. code-block:: javascript
 * 
 *      js> var city = new FEATURE.Feature({
 *        >     values: {
 *        >         location: new GEOM.Point([-110, 45]),
 *        >         name: "Metropolis"
 *        >     }
 *        > });
 *      js> city.get("name");
 *      Metropolis
 *      js> city.get("location");
 *      <Point [-110, 45]>
 */

exports.Feature = Feature;

// register a feature factory for the module
register(new Factory(Feature, {
    handles: function(config) {
        config = prepConfig(config);
        return (config.schema instanceof Schema);
    }
}));

File: src/main/java/org/geoscript/js/feature/Feature.java
package org.geoscript.js.feature;

import org.geoscript.js.GeoObject;
import org.geoscript.js.geom.Bounds;
import org.geoscript.js.geom.Geometry;
import org.geoscript.js.io.JSON;
import org.geoscript.js.proj.Projection;
import org.geotools.feature.simple.SimpleFeatureBuilder;
import org.mozilla.javascript.Context;
import org.mozilla.javascript.Function;
import org.mozilla.javascript.NativeObject;
import org.mozilla.javascript.ScriptRuntime;
import org.mozilla.javascript.Scriptable;
import org.mozilla.javascript.ScriptableObject;
import org.mozilla.javascript.Wrapper;
import org.mozilla.javascript.annotations.JSConstructor;
import org.mozilla.javascript.annotations.JSFunction;
import org.mozilla.javascript.annotations.JSGetter;
import org.mozilla.javascript.annotations.JSSetter;
import org.mozilla.javascript.annotations.JSStaticFunction;
import org.opengis.feature.GeometryAttribute;
import org.opengis.feature.Property;
import org.opengis.feature.simple.SimpleFeature;
import org.opengis.feature.simple.SimpleFeatureType;
import org.opengis.feature.type.AttributeDescriptor;
import org.opengis.feature.type.GeometryDescriptor;
import org.opengis.referencing.crs.CoordinateReferenceSystem;

public class Feature extends GeoObject implements Wrapper {

    /** serialVersionUID */
    private static final long serialVersionUID = 4426338466323185386L;
    
    /**
     * The GeoTools feature backing this feature.
     */
    private SimpleFeature feature;

    /**
     * Layer from which this feature was accessed.  Used to persist 
     * modifications.
     */
    Scriptable layer;

    /**
     * Prototype constructor.
     */
    public Feature() {
    }
    
    /**
     * Constructor with SimpleFeature (from Java).
     * @param scope
     * @param feature
     */
    public Feature(Scriptable scope, SimpleFeature feature) {
        this.feature = feature;
        setParentScope(scope);
        this.setPrototype(Module.getClassPrototype(Feature.class));
    }

    /**
     * Constructor from config object.
     * @param config
     */
    private Feature(NativeObject config) {
        Scriptable scope = ScriptableObject.getTopLevelScope(config);
        config = prepConfig(config);
        Schema schema = null;
        if (config.has("schema", config)) {
            Object schemaObj = config.get("schema", config);
            if (schemaObj instanceof Schema) {
                schema = (Schema) schemaObj;
            } else if (schemaObj instanceof Scriptable) {
                schema = new Schema(scope, (Scriptable) schemaObj);
            } else {
                throw ScriptRuntime.constructError("Error", "Cannot create schema from provided value: " + Context.toString(schemaObj));
            }
        }

        Object propertiesObj = getOptionalMember(config, "properties", NativeObject.class, "Object");
        NativeObject properties = null;
        if (propertiesObj != null) {
            properties = (NativeObject) propertiesObj;
        }

        if (schema == null) {
            if (properties != null) {
                schema = Schema.fromValues(scope, properties);
            } else {
                throw ScriptRuntime.constructError("Error", "Feature config must include schema or properties.");
            }
        }

        SimpleFeatureBuilder builder = new SimpleFeatureBuilder((SimpleFeatureType) schema.unwrap());

        if (properties != null) {
            Object[] names = properties.getIds();
            for (Object nameObj : names) {
                String name = (String) nameObj;
                if (schema.get(name) == null) {
                    throw ScriptRuntime.constructError("Error", "Feature schema has no field with the given name: " + name);
                }
                Object value = properties.get(name, properties);
                builder.set(name, jsToJava(value));
            }
        }
        
        String id = null;
        if (config.has("id", config)) {
            id = Context.toString(config.get("id", config));
        }

        feature = builder.buildFeature(id);
    }
    
    /**
     * Translate a GeoJSON config into a Feature config.
     * @param config
     * @return
     */
    private NativeObject prepConfig(NativeObject config) {
        Scriptable scope = config.getParentScope();
        Object propertiesObj = getOptionalMember(config, "properties", NativeObject.class, "Object");
        NativeObject properties = null;
        if (propertiesObj != null) {
            properties = (NativeObject) propertiesObj;
        }

        Geometry geometry = null;
        if (config.has("geometry", config)) {
            // GeoJSON config
            Object geometryObj = config.get("geometry", config);
            if (!(geometryObj instanceof NativeObject)) {
                throw ScriptRuntime.constructError("Error", 
                        "Expected geometry member to be an object, got: " + Context.toString(geometryObj));
            }
            geometryObj = JSON.readObj((NativeObject) geometryObj);
            if (!(geometryObj instanceof Geometry)) {
                throw ScriptRuntime.constructError("Error", "Expected geometry memeber to be a valid geometry, got: " + Context.toString(geometryObj));
            }
            geometry = (Geometry) geometryObj;
            config.delete("geometry");
            // allow passing a geometry only with no other properties
            if (properties == null) {
                Context cx = Context.enter();
                try {
                    properties = (NativeObject) cx.newObject(scope);
                } finally {
                    Context.exit();
                }
                config.put("properties", config, properties);
            }
        }
        if (geometry != null && properties != null) {
            properties.put("geometry", properties, geometry);
        }
        return config;
    }

    /**
     * Constructor from config object (without new keyword).
     * @param scope
     * @param config
     */
    public Feature(Scriptable scope, NativeObject config) {
        this(config);
        setParentScope(scope);
        this.setPrototype(Module.getClassPrototype(Feature.class));
    }

    @JSConstructor
    public static Object constructor(Context cx, Object[] args, Function ctorObj, boolean inNewExpr) {
        if (args.length != 1) {
            throw ScriptRuntime.constructError("Error", "Constructor takes a single argument");
        }
        Feature feature = null;
        Object arg = args[0];
        if (arg instanceof NativeObject) {
            NativeObject config = (NativeObject) arg;
            if (inNewExpr) {
                feature = new Feature(config);
            } else {
                feature = new Feature(config.getParentScope(), config);
            }
        } else {
            throw ScriptRuntime.constructError("Error", "Could not create feature from argument: " + Context.toString(arg));
        }
        return feature;
    }

    public Object unwrap() {
        return feature;
    }
    
    @JSGetter
    public String getId() {
        return feature.getIdentifier().toString();
    }
    
    @JSGetter
    public String getGeometryName() {
        GeometryAttribute geomProp = feature.getDefaultGeometryProperty();
        String name = null;
        if (geomProp != null) {
            name = geomProp.getName().toString();
        }
        return name;
    }
    
    @JSGetter
    public Geometry getGeometry() {
        String name = getGeometryName();
        return (Geometry) get(name);
    }
    
    @JSSetter
    public void setGeometry(Geometry geometry) {
        String name = getGeometryName();
        if (name == null) {
            throw ScriptRuntime.constructError("Error", "Feature schema has no geometry field");
        }
        set(name, geometry);
    }
    
    @JSGetter
    public Projection getProjection() {
        Projection projection = null;
        SimpleFeatureType featureType = feature.getFeatureType();
        GeometryDescriptor descriptor = featureType.getGeometryDescriptor();
        if (descriptor != null) {
            CoordinateReferenceSystem crs = descriptor.getCoordinateReferenceSystem();
            if (crs != null) {
                projection = new Projection(getParentScope(), crs);
            }
        }
        return projection;
    }
    
    @JSGetter
    public Bounds getBounds() {
        Bounds bounds = null;
        Geometry geometry = getGeometry();
        if (geometry != null) {
            bounds = geometry.getBounds();
        }
        return bounds;
    }

    @JSGetter
    public Schema getSchema() {
        return new Schema(getParentScope(), feature.getFeatureType());
    }

    @JSFunction
    public Feature set(String name, Object value) {
        SimpleFeatureType featureType = feature.getFeatureType();
        AttributeDescriptor descriptor = featureType.getDescriptor(name);
        if (descriptor == null) {
            throw ScriptRuntime.constructError("Error", "Feature schema has no such field: " + name);
        }
        String geomName = getGeometryName();
        if (geomName != null && geomName.equals(name)) {
            if (!(value instanceof Geometry)) {
                throw ScriptRuntime.constructError("Error", "Attempted to set geometry property to a non-geometry object: " + Context.toString(value));
            }
            Projection featureProj = getProjection();
            Geometry geometry = (Geometry) value;
            Projection geomProj = geometry.getProjection();
            if (featureProj != null) {
                if (geomProj != null) {
                    if (!featureProj.equals(geometry.getProjection())) {
                        value = geometry.transform(featureProj);
                    }
                }
            } else if (geomProj != null) {
                throw ScriptRuntime.constructError("Error", "Cannot add a geometry with a projection to a feature without a projection");
            }
        }
        feature.setAttribute(name, jsToJava(value));
        if (layer != null) {
            // queue modifications
            Object queueModifiedObj = ScriptableObject.getProperty(layer, "queueModified");
            if (!(queueModifiedObj instanceof Function)) {
                throw new RuntimeException("Unable to access queueModified method of layer.");
            }
            Function queueModified = (Function) queueModifiedObj;
            Context cx = getCurrentContext();
            Scriptable scope = getParentScope();
            Object[] args = { this, name };
            queueModified.call(cx, scope, layer, args);
        }
        return this;
    }

    @JSFunction
    public Object get(String name) {
        Object value;
        if (feature.getFeatureType().getDescriptor(name) == null) {
            value = Context.getUndefinedValue();
        } else {
            value = javaToJS(feature.getAttribute(name), getParentScope());
            if (value instanceof Geometry) {
                ((Geometry) value).setProjection(getProjection());
            }
        }
        return value;
    }

    /**
     * Set the layer from which this feature was accessed.
     * TODO: Remove this from the API when Layer is implemented as a wrapper.
     * @param layer
     * @return
     */
    @JSSetter
    public void setLayer(Scriptable layer) {
        if (this.layer != null) {
            // TODO: clear queued modifications?
        }
        this.layer = layer;
    }
    
    @JSGetter
    public Scriptable getLayer() {
        return layer;
    }

    
    @JSGetter
    public NativeObject getProperties() {
        Context cx = getCurrentContext();
        Scriptable scope = getParentScope();
        NativeObject properties = (NativeObject) cx.newObject(scope);
        for (Property property : feature.getProperties()) {
            String name = property.getName().toString();
            properties.put(name, properties, get(name));
        }
        return properties;
    }
    
    @JSSetter
    public void setProperties(Scriptable properties) {
        Object[] names = properties.getIds();
        for (Object nameObj : names) {
            String name = (String) nameObj;
            set(name, properties.get(name, properties));
        }
    }
    
    @JSGetter
    public Scriptable getConfig() {
        Scriptable config = super.getConfig();
        
        // add schema
        Schema schema = getSchema();
        Scriptable schemaConfig = schema.getConfig();
        config.put("schema", config, schemaConfig);

        // add default geometry member
        Geometry geometry = getGeometry();
        Scriptable geomConfig = null;
        if (geometry != null) {
            geomConfig = geometry.getConfig();
        }
        config.put("geometry", config, geomConfig);

        // add all other properties
        Context cx = getCurrentContext();
        Scriptable scope = getParentScope();
        Scriptable properties = cx.newObject(scope);
        Scriptable values = getProperties();
        String geometryName = getGeometryName();
        Object[] names = values.getIds();
        for (Object nameObj : names) {
            String name = (String) nameObj;
            if (!name.equals(geometryName)) {
                properties.put(name, properties, values.get(name, values));
            }
        }
        config.put("properties", config, properties);
        return config;
    }

    @JSStaticFunction
    public static Feature from_(Scriptable featureObj) {
        SimpleFeature feature = null;
        if (featureObj instanceof Wrapper) {
            Object obj = ((Wrapper) featureObj).unwrap();
            if (obj instanceof SimpleFeature) {
                feature = (SimpleFeature) obj;
            }
        }
        if (feature == null) {
            throw ScriptRuntime.constructError("Error", "Cannot create feature from " + Context.toString(featureObj));
        }
        return new Feature(getTopLevelScope(featureObj), feature);
    }

    public String toFullString() {
        NativeObject properties = getProperties();
        Object[] names = properties.getIds();
        String repr = "";
        int length = names.length;
        for (int i=0; i<length; ++i) {
            String name = (String) names[i];
            Object value = get(name);
            if (value instanceof Geometry) {
                value = "<" + value.getClass().getSimpleName() + ">";
            } else if (value instanceof String) {
                value = '"' + (String) value + '"';
            } else {
                value = Context.toString(value);
            }
            repr += name + ": " + value;
            if (i < length - 1) {
                repr += ", ";
            }
        }
        return repr;
    }


}
